source/postprocess/particle_distribution_score.cc

```diff
@@ -0,0 +1,282 @@
+    }
+
+
+
+    template <int dim>
+    std::list<std::string>
+    ParticleDistributionScore<dim>::required_other_postprocessors() const
+    {
+      return {"particles"};
+    }
+
+
+
+    template <int dim>
+    void ParticleDistributionScore<dim>::sort_particles_into_buckets(
+      const typename Triangulation<dim>::active_cell_iterator &cell,
+      const double bucket_width,
+      Table<dim,unsigned int> &buckets) const
+    {
+      for (unsigned int particle_manager_index = 0; particle_manager_index < this->n_particle_managers(); ++particle_manager_index)
+        {
+          // sort the particles within the current cell
+          const typename Particle::ParticleHandler<dim>::particle_iterator_range particle_range =
+            this->get_particle_manager(particle_manager_index).get_particle_handler().particles_in_cell(cell);
+
+          for (const auto &particle: particle_range)
+            {
+              const double particle_x = particle.get_reference_location()[0];
+              const double particle_y = particle.get_reference_location()[1];
+
+              const double x_ratio = (particle_x) / (bucket_width);
+              const double y_ratio = (particle_y) / (bucket_width);
+
+              unsigned int x_index = static_cast<unsigned int>(std::floor(x_ratio));
+              unsigned int y_index = static_cast<unsigned int>(std::floor(y_ratio));
+
+              /*
+              If a particle is exactly on the boundary of two cells its
+              reference location will equal 1, and if this is the case,
+              the "x/y/z_index" will be outside of the range of the table without
+              these checks. The table has a number of entries equal to "granularity" in each dimension,
+              and the table is indexed at 0, so if the "x/y/z_indez" equals "granularity" it
+              will be out of range.
+              */
+              if (x_index == granularity)
+                x_index = granularity-1;
+              if (y_index == granularity)
+                y_index = granularity-1;
+
+              TableIndices<dim> entry_index;
+              entry_index[0] = x_index;
+              entry_index[1] = y_index;
+              if (dim == 3)
+                {
+                  const double particle_z = particle.get_reference_location()[2];
+                  const double z_ratio = (particle_z) / (bucket_width);
+                  unsigned int z_index = static_cast<unsigned int>(std::floor(z_ratio));
+                  if (z_index == granularity)
+                    z_index = granularity-1;
+                  entry_index[2] = z_index;
+                }
+
+              ++buckets(entry_index);
+            }
+        }
+    }
+
+
+
+    template <int dim>
+    void
+    ParticleDistributionScore<dim>::declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Postprocess");
+      {
+        prm.enter_subsection("Particle Distribution Score");
```

Comment:
I probably didnt explain this before, but our usual style for subsection naming is this (unless the name contains proper nouns):
```suggestion
        prm.enter_subsection("Particle distribution score");
```